samples/SampleMsmqHost/MsmqConnection.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.IO;
+using System.Messaging;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace SampleMsmqHost
+{
+    public interface IMsmqConnection
+    {
+        Task SendTextAsync(string text, CancellationToken cancellationToken);
+
+        Task<Message> ReceiveAsync(CancellationToken cancellationToken);
+    }
+
+    public class MsmqConnection : IMsmqConnection, IDisposable
+    {
+        private readonly MessageQueue _queue;
+
+        public MsmqOptions Options { get; }
+
+        public ILogger<MsmqConnection> Logger { get; }
+
+        public MsmqConnection(IOptions<MsmqOptions> options, ILogger<MsmqConnection> logger)
+        {
+            Options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+            Logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _queue = OpenQueue();
+        }
+
+        private MessageQueue OpenQueue()
+        {
+            Logger.LogInformation("Opening Queue: Path={0}; AccessMode={1};", Options.Path, Options.AccessMode);
+
+            return new MessageQueue(Options.Path, Options.SharedModeDenyReceive, Options.EnableCache, Options.AccessMode);
+        }
+
+        public void Dispose()
+        {
+            Logger.LogInformation("Closing Queue");
+
+            _queue?.Dispose();
+        }
+
+        public async Task SendTextAsync(string text, CancellationToken cancellationToken)
```

Comment:
Fake async?

Reply:
Sorry, when I first wrote the interface signature, I assumed that `MessageQueue` had an APM version of `Send` but later realized that it didn't. Code updated to be synchronous now.